client/src/Attendance/Attendance.js

```diff
@@ -0,0 +1,281 @@
+      (value, index) => {
+        // this logic adds and removes fields as needed depending on whether checkin/out or an absence is selected
+        return index === i
+          ? Object.keys(updates).length === 0 ||
+            (Object.keys(value).includes('absence') &&
+              (Object.keys(updates).includes('check_in') ||
+                Object.keys(updates).includes('check_out'))) ||
+            (Object.keys(updates).includes('absence') &&
+              (Object.keys(value).includes('check_in') ||
+                Object.keys(value).includes('check_out')))
+            ? updates
+            : { ...value, ...updates }
+          : value
+      }
+    )
+    latestAttendanceData.current = {
+      ...attendanceData,
+      [record.id]: newArr
+    }
+    setAttendanceData(prevData => ({ ...prevData, [record.id]: newArr }))
+  }
+
+  const handleDateChange = (index, ds) => {
+    const updatedDates = latestColumnDates.current.map((value, i) =>
+      index === i ? ds : value
+    )
+    latestColumnDates.current = updatedDates
+    setColumnDates(updatedDates)
+  }
+
+  const generateColumns = () => {
+    let cols = []
+    for (let i = 0; i < 7; i++) {
+      cols.push({
+        dataIndex: 'date' + i,
+        key: 'date' + i,
+        width: 398,
+        // eslint-disable-next-line react/display-name
+        title: () => (
+          <DatePicker
+            disabledDate={c => c && c.valueOf() > Date.now()}
+            onChange={(_, ds) => handleDateChange(i, ds)}
+            bordered={false}
+            placeholder={t('selectDate')}
+            style={{ width: '8rem ', color: '#004A6E' }}
+          />
+        ),
+        // eslint-disable-next-line react/display-name
+        render: (_, record) => {
+          return (
+            <AttendanceDataCell
+              record={record}
+              columnIndex={i}
+              updateAttendanceData={updateAttendanceData}
+            />
+          )
+        }
+      })
+    }
+
+    return [
+      {
+        title: t('childNameCap'),
+        dataIndex: 'name',
+        width: 250,
+        key: 'name',
+        // eslint-disable-next-line react/display-name
+        render: (_, record) => {
+          return (
+            <div>
+              <p className="text-lg mb-1">
+                {record.childName || record.child.childName}
+              </p>
+              <p className="flex flex-wrap mt-0.5">
+                {record.business || record.child.business}{' '}
+                <img className="mx-1" alt="ellipse" src={ellipse} />{' '}
+                {record.cNumber || record.child.cNumber}
+              </p>
+            </div>
+          )
+        }
+      },
+      ...cols
+    ]
+  }
+
+  const [columns, setColumns] = useState(generateColumns())
+
+  i18n.on('languageChanged', () => setColumns(generateColumns()))
+
+  const handleSave = async () => {
+    const attendanceBatch = Object.entries(attendanceData).flatMap(cv =>
```

Comment:
I know this is sort of small but can we use words instead of abbreviations for variables like this and v/k for hash maps?  I spent like 45 seconds trying to figure out why we were mapping a CSV before I realized I'd misread it and it probably meant `columnValues` 😂

Reply:
yes!